src/server/frontend_xwayland/xwayland_surface.cpp

```diff
@@ -1047,3 +966,144 @@ auto mf::XWaylandSurface::latest_input_timestamp(std::lock_guard<std::mutex> con
         return {};
     }
 }
+
+void mf::XWaylandSurface::apply_any_mods_to_scene_surface()
+{
+    std::shared_ptr<mir::scene::Surface> scene_surface;
+    std::experimental::fundamentals_v1::optional<std::unique_ptr<mir::shell::SurfaceSpecification>> spec;
+
+    {
+        std::lock_guard<std::mutex> lock{mutex};
+        if ((scene_surface = weak_scene_surface.lock()))
+        {
+            spec = consume_pending_spec(lock);
+        }
+    }
+
+    if (spec && scene_surface)
+    {
+        if (spec.value()->application_id.is_set() &&
+            spec.value()->application_id.value() == scene_surface->application_id())
+            spec.value()->application_id.consume();
+
+        if (spec.value()->name.is_set() &&
+            spec.value()->name.value() == scene_surface->name())
+            spec.value()->name.consume();
+
+        if (spec.value()->parent.is_set() &&
+            spec.value()->parent.value().lock() == scene_surface->parent())
+            spec.value()->parent.consume();
+
+        if (!spec.value()->is_empty())
+            shell->modify_surface(scene_surface->session().lock(), scene_surface, *spec.value());
+    }
+}
+
+void mf::XWaylandSurface::window_type(xcb_atom_t wm_type)
+{
+    auto set_type = [this](MirWindowType type)
+        {
+            std::lock_guard<std::mutex> lock{mutex};
+            pending_spec(lock).type = type;
+        };
+
+    if (this->connection->net_wm_window_type_normal == wm_type)
+    {
+        set_type(mir_window_type_freestyle);
+    }
+    else if (this->connection->net_wm_window_type_popup == wm_type)
+    {
+        set_type(mir_window_type_gloss);
+    }
+    else if (this->connection->net_wm_window_type_menu == wm_type)
+    {
+        set_type(mir_window_type_menu);
+    }
+    else
+    {
+        set_type(mir_window_type_freestyle);
+
+        if (verbose_xwayland_logging_enabled())
+        {
+            log_debug("Ignoring type (%s) of %s",
+                      this->connection->query_name(wm_type).c_str(),
+                      this->connection->window_debug_string(window).c_str());
+        }
+    }
+}
+
+void mf::XWaylandSurface::set_parent(xcb_window_t xcb_window, std::lock_guard<std::mutex> const& lock)
+{
```

Comment:
I think the error conditions should explicitly clear the parent (or the parent should be cleared at the start of the function, and stay cleared unless it is set)